tests/unit/wait-for-coroutine-test.ts

```diff
@@ -0,0 +1,113 @@
+import { task } from 'ember-concurrency-decorators';
+import EmberObject from '@ember/object';
+import { task as taskFn } from 'ember-concurrency';
+import { DEBUG } from '@glimmer/env';
+
+if (DEBUG) {
+  module('wait-for-coroutine', function(hooks) {
+    hooks.afterEach(function() {
+      _reset();
+      resetError();
+    });
+
+    class DecoratorThing {
+      @task
+      @waitForCoroutine
+      *doAsyncStuff(...args: any) {
+        yield new Promise(resolve => {
+          resolve();
+        });
+        return [this, ...args];
+      }
+
+      @task
+      @waitForCoroutine
+      *asyncThrow() {
+        yield new Promise(resolve => {
+          resolve();
+        });
+        throw new Error('doh!');
+      }
+    }
+
+    class FunctionThing extends EmberObject.extend({
+      doAsyncStuff: taskFn(
+        waitForCoroutine(function* doAsyncStuff(this: FunctionThing, ...args: any) {
+          yield new Promise(resolve => {
+            resolve();
+          });
+          return [this, ...args];
+        })
+      ),
+
+      asyncThrow: taskFn(
+        waitForCoroutine(function* asyncThrow() {
+          yield new Promise(resolve => {
+            resolve();
+          });
+          throw new Error('doh!');
+        })
+      ),
+    }) {}
+
+    let thing: DecoratorThing | FunctionThing;
+
+    function runTests() {
+      test('waitForCoroutine wraps and registers a waiter', async function(assert) {
+        overrideError(MockStableError);
+
+        // @ts-ignore
+        let promise: Promise<{}> = thing.doAsyncStuff.perform();
+
+        assert.deepEqual(getPendingWaiterState(), {
+          pending: 1,
+          waiters: {
+            'coroutine-waiter': [
+              {
+                label: undefined,
+                stack: 'STACK',
+              },
+            ],
+          },
+        });
+
+        await promise.then(() => {
+          assert.deepEqual(getPendingWaiterState(), { pending: 0, waiters: {} });
+        });
+      });
+
+      test('waitForCoroutine handles arguments and return value', async function(assert) {
+        overrideError(MockStableError);
+
+        // @ts-ignore
+        let ret = await thing.doAsyncStuff.perform(1, 'foo');
+        assert.deepEqual(ret, [thing, 1, 'foo']);
+      });
+
+      test('waitForCoroutine transitions waiter to not pending even if coroutine throws', async function(assert) {
+        // @ts-ignore
+        await thing.asyncThrow.perform().catch(() => null);
+        assert.deepEqual(getPendingWaiterState(), { pending: 0, waiters: {} });
+      });
+    }
+
+    module('decorator', function(hooks) {
+      hooks.beforeEach(function() {
+        thing = new DecoratorThing();
+      });
+
+      runTests();
+    });
+
+    module('function', function(hooks) {
+      hooks.beforeEach(function() {
+        thing = FunctionThing.create();
+      });
+
+      runTests();
+    });
```

Comment:
I'd opt to pass a module name and type in as configuration in a loop, similar to how we did in ember-lifeline https://github.com/ember-lifeline/ember-lifeline/blob/3e331f97fa6a537a71993a918dc10900137944ab/tests/unit/dom-event-listeners-test.js#L43-L52

I guess I find it slightly easier to read if we do that.

Reply:
I could do that, but since they have different setup steps (`beforeEach()`), it would be a bit more convoluted than the `ember-lifeline` example, since it would involve passing in a function with the `beforeEach` logic. IMO this reads more cleanly, but I'm sure it's stylistic, and this isn't my project, so if you still prefer the other way I'm happy to switch.

Reply:
This is definitely in the realm of 'overly pedantic' in terms of should we do this vs. the structure in ember-lifeline. I'm inclined to say just ignore this, as I can go either way. 

It struck me that it took me a minute to orient myself on how the tests were structured, which is the reason for my comment. This isn't a make or break thing for me, unlike the choice of using mocha/chai above which is :)

Reply:
👍 Will definitely convert to QUnit once we've collected/processed the feedback and it's time for another coding pass. I chose mocha because I'm much more used to seeing Node tests written in mocha (although I generally prefer QUnit myself) and thought that would keep it more familiar to people accustomed to Node and babel plugin testing and whatnot, but watch how eagerly I ditch it and replace with QUnit 😃